fairscale/nn/data_parallel/sharded_ddp.py

```diff
@@ -4,109 +4,178 @@
+
+        # Expose the same attributes as PytorchDDP, some frameworks rely on them.
+        # See https://pytorch.org/docs/stable/_modules/torch/nn/parallel/distributed.html#DistributedDataParallel
+        self.is_multi_device_module = len({p.device for p in self.base_model.parameters()}) > 1
+        distinct_device_types = {p.device.type for p in self.base_model.parameters()}
+        assert len(distinct_device_types) == 1, (
+            "ShardedDataParallel's input module must be on "
+            "the same type of devices, but input module parameters locate in {}."
+        ).format(distinct_device_types)
+        self.device_type = list(distinct_device_types)[0]
+
+        # Fill in a look-up table per grad
+        self._grad_to_rank = [
+            _get_global_rank(self.process_group, self._find_rank(p)[1])
+            for p in filter(lambda x: x.requires_grad, self.base_model.parameters())
+        ]
+
+        # Setup the reduce buckets
+        # reduce_buffer dimensions: [optimizer, device, rank params]
+        # bucket_strategy dimensions: [optimizer, device, rank tuple -should bucket, offset_start, offset_end-]
+        # bucket_state dimensions: [optimizer, device, rank - current offset, max offset-]
+        self._reduce_buckets: Dict[OSS, Dict[torch.device, List[Dict[str, Any]]]] = {}
+        self._bucket_strategy: List[Tuple[bool, int, int]] = []
+        self._bucket_state: Dict[OSS, Dict[torch.device, List[Tuple[int, int]]]] = {}
+        self._buffer_size = min(reduce_buffer_size, sum(p.numel() for p in self.base_model.parameters()))
+        self._setup_buckets(reduce_buffer_size)
+
+        # Scafolding to be able to reduce the grads during the BW pass
+        self._grad_to_be_reduced = [True for _ in filter(lambda x: x.requires_grad, self.base_model.parameters())]
+        self._grad_accs: List[Callable] = []
+        self._reduce_work_handles: List[Any] = []
+        self._setup_backward_hooks()
+
+        # Make sure that all ranks start with the same model
+        self.sync_all_params()
+
+    def forward(self, *inputs: Any, **kwargs: Any) -> Any:
+        if self.broadcast_buffers:
+            self.sync_buffers()
+
+        # Reset all the grad reduce and bucket state flags
+        self._grad_to_be_reduced = [True for _ in self._grad_to_be_reduced]
+        for sharded_optimizer in self.sharded_optimizers:
+            for device, per_rank_params in sharded_optimizer.per_device_params.items():
+                for r in range(self.world_size):
+                    self._bucket_state[sharded_optimizer][device][r] = (
+                        0,
+                        self._bucket_state[sharded_optimizer][device][r][1],
+                    )
 
-        # Flag used to make sure we only reduce gradients one time in the execution engine
-        self.need_reduction = False
+        # mark tensors as "requires_grad" to register the Gatekeeper in the autograd graph
```

Comment:
This will not work for non-float input tensors. The inputs in NLP are not floats.

Can you add a test for the non-float case.

Reply:
> This is very cool. Thanks for doing this heroic work! I am thinking of several things to do in the future. Do you mind I file them as issues?
> 
> 1. if the buckets are currently the same size as the grads, perhaps we should introduce a mode where a fixed sized buckets are used to reduce memory consumption?

the buckets are fixed size right now, it's a parameter actually. They are a lot smaller than the grads, it's only to collect the smallest ones, most of them are reduced directly. Two follow ups that I see are that we could use tensor views to remove the need for a copy, and/or reuse the buckets that OSS also has, they should be the same

> 2. we should make sure this works without OSS?

I don't see how that is possible, the whole point is that it reduces the gradients to the right ranks, you need a sharded optimizer for that ?

> 3. after 1, to avoid python GIL overhead, perhaps we should move reducing into its own python process and use queues to send reduction data to that process for parallel reduction?
ah yes, it would be nice to measure that, there's probably some perf to win there, I agree

> 4. ensure this works with checkpointing
yes, good point ! there's a context manager to accumulate/block the reduce, so I'm guessing that it should work already, but would be nice to test

Reply:
Fixed, new take which removes this faulty requirement